KIPs/kip-146.md

```diff
@@ -0,0 +1,157 @@
+
+Parameters:
+
+- `validators`: the deterministically sorted list of validators.
+- `round`: the target consensus round.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` (from [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md)) of the previous block is used (i.e., previous to the target block).
+
+```py
+def proposer_selector(validators, round, seed):
+    return shuffle_validators_KIP146(validators, round, seed)[round % len(validators)]
+```
+
+### Committee selection
+
+For calculating the committee of `FORK_BLOCK` and afterward, the new logic must be used.
+This only affects the chain whose proposer selection policy is set to `WeightedRandom`.
+
+This design guarantees that the given proposer address is always in the committee.
+
+Parameters:
+
+- `validators`: the deterministically sorted list of validators.
+- `commitee_size`: the value of `istanbul.sub` in ChainConfig.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer. The `mixHash` (from [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md)) of the previous block is used (i.e., previous to the target block).
+- `round`: the target consensus round.
+
+```py
+def select_committee_KIP146(validators, committee_size, seed, round):
+    shuffled = shuffle_validators_KIP146(validators, round, seed)
+    return (shuffled[round:] + shuffled[:round])[:committee_size]
+
+def compute_committee(validators, committee_size, proposer, round, seed):
+    committee = select_committee_KIP146(validators, committee_size, seed, round)
+    if proposer not in committee:
+        committee = committee[:-1] + proposer
+    return committee
+```
+
+## Rationale
+
+### Using part of `mixHash` as a seed
+
+As defined in [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md), `mixHash` is the result of XOR operation with a keccak256 hash.
+To the best of our knowledge, truncating the cryptographic hash is safe to be used for the seed of pseudo random number generator.
+Ethereum also uses the first eight bytes of a hash: [link](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#compute_shuffled_index).
+
+### Side effect
+
+The current algorithm for composing a committee makes sure that the proposer and the next round proposer is included in the committee. Before `FORK_BLOCK`, the next round proposer is equal to the next block proposer, and thus the next block proposer always ended up in the committee. However, after `FORK_BLOCK`, the next round proposer can be different from the next block proposer, which may result in the next block proposer not being included from the committee. This side effect does not degrade the network security; while the next proposer being in the committee could potentially be helpful to next block generation, its significance is negligible, if any.
```

Comment:
I agree with this point, but it would be good to check the performance with a heavy network load.